op-challenger/game/fault/trace/cannon/provider.go

```diff
@@ -56,12 +55,7 @@ type CannonTraceProvider struct {
 	lastStep uint64
 }
 
-func NewTraceProvider(ctx context.Context, logger log.Logger, m CannonMetricer, cfg *config.Config, gameContract *contracts.FaultDisputeGameContract, localContext common.Hash, dir string, gameDepth uint64) (*CannonTraceProvider, error) {
-	l2Client, err := ethclient.DialContext(ctx, cfg.CannonL2)
-	if err != nil {
-		return nil, fmt.Errorf("dial l2 client %v: %w", cfg.CannonL2, err)
-	}
-	defer l2Client.Close() // Not needed after fetching the inputs
+func NewTraceProvider(ctx context.Context, logger log.Logger, m CannonMetricer, cfg *config.Config, l2Client L2HeaderSource, gameContract *contracts.FaultDisputeGameContract, localContext common.Hash, dir string, gameDepth uint64) (*CannonTraceProvider, error) {
```

Comment:
hmm, my thinking here was that we would want an L2 client that lives for the entire lifetime of the challenger in the same way we have a single L1 client for the entire challenger. The complication is that the l2 client is only needed if cannon is used in some way and the L2 URL is only required in config in those cases.

We can't make the functions private in this case because they are used by e2e helpers as well and it's quite helpful to minimise the amount of setup code we have to duplicate in e2e, though it wouldn't be the end of the world in this case.

Possibly the best option is to just have a single `NewTraceProvider` that takes the `LocalGameInputs` and doesn't need the L2 client, then just expose the couple of methods to fetch local game inputs for the required situations and reuse those in e2e. I'll have a play with that.

Reply:
Exposing the fetch functions and always passing in the retrieved local inputs works out well - simpler all round and avoids the potential lifecycle confusion you mention.

Reply:
Nice!